p2p/src/peers/mod.rs

```diff
@@ -1,85 +1,271 @@
 //! Library for managing a list of available peers
 
-use serde::{Deserialize, Serialize};
-
-use std::collections::HashMap;
-use std::net::SocketAddr;
-
 use rand::{thread_rng, Rng};
+use serde::{Deserialize, Serialize};
+use std::{collections::HashMap, net::SocketAddr, u8};
 
+use witnet_crypto::hash::calculate_sha256;
 use witnet_util::timestamp::get_timestamp;
 
 /// Peer information being used while listing available Witnet peers
 #[derive(Serialize, Deserialize)]
 struct PeerInfo {
     address: SocketAddr,
-    _timestamp: i64,
+    timestamp: i64,
 }
 
 /// Peers TBD
 #[derive(Default, Serialize, Deserialize)]
 pub struct Peers {
-    /// Server sessions
-    peers: HashMap<SocketAddr, PeerInfo>,
+    /// Bucket for tried addresses
+    tried_bucket: HashMap<u16, PeerInfo>,
+    /// Bucket for new addresses
+    new_bucket: HashMap<u16, PeerInfo>,
+    /// Nonce value
+    sk: u8,
 }
 
 impl Peers {
-    /// Add multiple peer addresses and save timestamp
+    /// Create a new instance of Peers
+    pub fn new() -> Self {
+        Peers {
+            sk: thread_rng().gen_range(0, u8::MAX),
+            tried_bucket: HashMap::new(),
+            new_bucket: HashMap::new(),
+        }
+    }
+
+    /// Algorithm to calculate index for the tried addresses buckets
+    fn tried_bucket_index(&self, socket_addr: &SocketAddr) -> u16 {
+        let (ip, group, host_id) = split_socket_addresses(socket_addr);
+
+        calculate_index_for_tried(self.sk, &ip, &group, &host_id)
+    }
+
+    /// Algorithm to calculate index for the new addresses buckets
+    fn new_bucket_index(&self, socket_addr: &SocketAddr, src_socket_addr: &SocketAddr) -> u16 {
+        let (_, group, host_id) = split_socket_addresses(socket_addr);
+        let (_, src_group, _) = split_socket_addresses(src_socket_addr);
+
+        calculate_index_for_new(self.sk, &src_group, &group, &host_id)
```

Comment:
I see that the `index_for_new` depends on the source socket address: the address of the peers that sent us the list of peers. What happens when two different peers send us the same peer? Will it be stored twice in different buckets? Or should be check if we already had this peer in new. And what about if we already have this peer in tried?

Reply:
If 2 different peers send us the same peer will be saved in a different bucket. @girazoki explained me that this feature is important to have redundancy of good peers against possible peers provided by a bad peer. Also, if you do not know the other peer sources is impossible to check if this peer exist in new. Respect to check in tried, @girazoki say that the new and tried buckets has to be independents, so we do not need to check if it is in tried